database/utils.py

```diff
@@ -87,48 +64,46 @@ def cleanup():
     lock = None
 
 
-@use_session
-def add_all(entities):
-    """Save all |entities| to the database connected to by session."""
+@contextmanager
+def session_scope():
+    """Provide a transactional scope around a series of operations."""
+    global session
+    global engine
+    global lock
+    if session is None or engine is None or lock is None:
+        initialize()
+    lock.acquire()
     try:
-        session.add_all(entities)
-        session.commit()
-        return
+        yield session
     except Exception as e:
         session.rollback()
         raise e
+    finally:
+        lock.release()
 
 
-@use_session
-def bulk_save(entities):
+def add_all(entities):
     """Save all |entities| to the database connected to by session."""
-    try:
-        session.bulk_save_objects(entities)
-        session.commit()
-        return
-    except Exception as e:
-        session.rollback()
```

Comment:
I'm somewhat worried we are getting rid of all of these rollbacks. Why are doing this?

Reply:
Also, we are not fixing global session problem atm, right now it is trying to fix the missing rollbacks first. so session_scope gives all the missing rollbacks.

Reply:
Ah sorry misread this PR then.
`session_scope` isn't the best name then for what it currently does, but it's fine I guess.

Reply:
in next pr, i can do the actual throwing away of session in session scope, right now trying to stabilize things first, esp for 7 day exp.